### PR TITLE
feat(fast-api): code +  test + ci 

### PR DESCRIPTION
**Pull request by Sanix-Darker**
## WHAT 

- deps: added missing dependencies for tests
- test: added meta fixtures for tests
- feat: some class compositions + inheritances
- updates and cleaning on test
- proper fixture for the database

taking updates from  : https://github.com/osscameroon/onacut/pull/150

### CODE REVIEW

File: fastapi/onacut/tests/routers/test_cities.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import pytest
 
-from . import client
+from onacut.tests import client
```
Comment (Zaker237) [inline fastapi/onacut/tests/routers/test_cities.py:4]: Nice here thanks

File: backend/onacut.sql [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	region_id INTEGER, 
+	longitude FLOAT, 
+	lattitude FLOAT, 
+	city_id INTEGER, 
+	district_id INTEGER, 
+	PRIMARY KEY (id), 
+	FOREIGN KEY(region_id) REFERENCES region (id), 
+	FOREIGN KEY(city_id) REFERENCES city (id), 
+	FOREIGN KEY(district_id) REFERENCES district (id)
+);
```
Comment (Zaker237) [inline backend/onacut.sql:79]: Good idea

File: .github/workflows/fastapi-ci.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      - .github/workflows/fastapi-ci.yaml
+      - fastapi/**
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ['3.10']
```
Comment (Zaker237) [inline .github/workflows/fastapi-ci.yaml:15]: why the 3.10 version??

**Review by Zaker237 — APPROVED** [review]

File: fastapi/onacut/schemas/region.py [hunk 4]
```diff
@@ -8,7 +8,7 @@ class RegionBase(BaseModel):
 
 
 class RegionCreate(RegionBase):
-    pass
+    pass # why not just use the RegionBase if this class is empty ?
```
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply fastapi/onacut/schemas/region.py]: mhhh... i see

File: fastapi/onacut/tests/routers/test_alerts.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 import pytest
+
 from . import client
```
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply fastapi/onacut/tests/routers/test_alerts.py]: this is fixed everywhere 🤔  on future commits

File: .github/workflows/fastapi-ci.yaml (hunk 3 above)
Comment (Sanix-Darker) [reply .github/workflows/fastapi-ci.yaml:15]: for perfs i guess...  and the typing there is not strong constraints !

File: fastapi/onacut/tests/routers/test_alerts.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zaker237) [reply fastapi/onacut/tests/routers/test_alerts.py]: alright!

File: .github/workflows/fastapi-ci.yaml (hunk 3 above)
Comment (Zaker237) [reply .github/workflows/fastapi-ci.yaml:15]: oki nice

File: fastapi/onacut/tests/routers/test_cities.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        "lattitude": 4.012475,
+    }
+
+
+def test_read_all_cities() -> None:
+    response = client.get("/cities")
+    assert response.status_code == 200
+
+
+def test_add_city(random_name: str, metadata: dict, region_id: int) -> None:
```
Comment (elhmn) [inline fastapi/onacut/tests/routers/test_cities.py:19]: out of curiosity, what is calling this test functions and setting this arguments ?

File: fastapi/onacut/tests/routers/conftest.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+    engine = create_engine(
+        os.path.join("sqlite:///onacut/tests/fixtures/test_sqlite.data"),
+        connect_args={"check_same_thread": False},
+    )
+
+    # we need to have a fresh looking test db, so for all our tables we're
+    # going to delete everything
+    with contextlib.closing(engine.connect()) as con:
+        trans = con.begin()
```
Comment (elhmn) [inline fastapi/onacut/tests/routers/conftest.py:41]: what does `trans` stand for ?

File: fastapi/onacut/tests/routers/conftest.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        **{
+            "type": "electricity",
+            "region_id": region_id,
+            "city_id": city_id,
+            "district_id": district_id,
+        },
+    }
+    response = client.post("/alerts/", json=alert)
+
+    return response.json()["id"]
```
Comment (elhmn) [inline fastapi/onacut/tests/routers/conftest.py:100]: This confuses me, the test code is supposed to test that the client work but for some reason we are already using the client to setup fixtures. If the client is broken the fixture setup will be broken which will make our test irrelevant.

File: fastapi/onacut/tests/routers/test_cities.py (hunk 6 above)
Comment (elhmn) [reply fastapi/onacut/tests/routers/test_cities.py:19]: Uhm are they coming from the `config_test` fixture setup ?

Comment (Sanix-Darker) [reply fastapi/onacut/tests/routers/test_cities.py:19]: exactly !

File: fastapi/onacut/tests/test_common.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        "test",
+        234,
+        {
+            "this": {
+                "oups": {"yy": {"value3": 0.99}},
+                "test": {"value1": "welcome", "value2": YY(value3=0.99)},
+            }
+        },
+        {"value3": 0.99},
+    ]
```
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply fastapi/onacut/tests/test_common.py]: 👍🏽

File: fastapi/onacut/tests/routers/test_alerts.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+            if region == _BAD_ID
+            else ("city" if city == _BAD_ID else "district")
+        )
+        assert response.json() == {"detail": f"Bad {bad_element}'s id!"}
+
+
+def test_delete_alert(alert_id: int) -> None:
+    """Should delete an alert"""
+    res = client.delete(f"/alerts/{alert_id}")
+    assert res.status_code == 200
```
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply fastapi/onacut/tests/routers/test_alerts.py:78]: 👍🏽

File: fastapi/onacut/tests/routers/test_alerts.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        response = client.post("/alerts/", json=alert)
+
+        assert response.status_code == 400
+        bad_element = (
+            "region"
+            if region == _BAD_ID
+            else ("city" if city == _BAD_ID else "district")
+        )
```
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply fastapi/onacut/tests/routers/test_alerts.py]: 👍🏽

File: fastapi/onacut/tests/routers/test_alerts.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    assert item == alert
+
+
+def test_add_alert_bad_region_city_district(
+    metadata: dict, region_id: str, city_id: int, district_id: int
+) -> None:
+    for region, city, district in [
+        (_BAD_ID, city_id, district_id),
+        (region_id, _BAD_ID, district_id),
+        (region_id, city_id, _BAD_ID),
```
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply fastapi/onacut/tests/routers/test_alerts.py]: true !
updated !

File: fastapi/onacut/tests/routers/test_alerts.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            "city_id": city_id,
+            "district_id": district_id,
+        },
+    }
+
+    response = client.post("/alerts/", json=alert)
+
+    assert response.status_code == 200
+    item = response.json()
+    item.pop("id", None)  # delete the id from the response
```
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply fastapi/onacut/tests/routers/test_alerts.py:41]: the id is the returned `id` from the sqlLite, i think this part is ok for now and was already present on the @Zaker237 PR sooo...

File: fastapi/onacut/tests/routers/test_alerts.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+@pytest.fixture
+def metadata() -> dict:
+    return {
+        "observations": "Test on Alert",
+        "date": "2022-10-30",
+        "begin_time": "18:00:00",
+        "longitude": 11.124576,
+        "lattitude": 4.012475,
+        "end_time": "20:30:00",
+    }
```
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply fastapi/onacut/tests/routers/test_alerts.py:15]: those are juste mock test no need to be precise, it's just to fit the model when adding a new resource !

File: .github/workflows/fastapi-ci.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply .github/workflows/fastapi-ci.yaml]: 👍🏽

**Review by elhmn — APPROVED** [review]

**Pull request merged**